### PR TITLE
5 leaves per node for the vote option tree

### PR DESCRIPTION
**Pull request by weijiekoh**
[These benchmarks on Poseidon quinary Merkle trees](https://ethresear.ch/t/gas-and-circuit-constraint-benchmarks-of-binary-and-quinary-incremental-merkle-trees-using-the-poseidon-hash-function/7446) show that we can have higher-capacity vote option trees with significantly fewer circuit constraints. The state and message trees, however, should stick to using Poseidon binary Merkle trees or users will have to pay higher gas fees for signups and message publishing.

This PR replaces the Merkle tree class in `maci-crypto` with `IncrementalQuinTree`, which supports both binary and quinary incremental Merkle trees. Due to a bug it does not support 3 or 4 leaves per node yet.

There are also new quinary Merkle tree circuits which support leaf existence checks, path verification, and root computation.

### CODE REVIEW

File: circuits/circom/trees/incrementalQuinTree.circom [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    signal input in; 
+    signal output out[levels];
+
+    var m = in;
+    for (var i = 0; i < levels; i ++) {
+        // circom's best practices state that we should avoid using <-- unless
+        // we know what we are doing. But this is the only way to perform the
+        // modulo operation. Later, we will do a range proof to ensure that the
+        // output is correct.
+        out[i] <-- m % BASE;
```
Comment (weijiekoh) [inline circuits/circom/trees/incrementalQuinTree.circom:231]: @kobigurk @ChihChengLiang `<--` is usually a red flag for circom circuits, so I'm tagging you to seek your advice on whether the range proof below is sufficient :smile:

Comment (kobigurk) [reply circuits/circom/trees/incrementalQuinTree.circom:231]: LGTM!

File: circuits/circom/updateStateTree.circom [hunk 2]
```diff
@@ -251,7 +254,7 @@ template UpdateStateTree(
 
     // Vote options tree root (supplied by coordinator)
     signal private input vote_options_tree_root;
-    signal private input vote_options_tree_path_elements[vote_options_tree_depth];
+    signal private input vote_options_tree_path_elements[vote_options_tree_depth][4];
```
Comment (kobigurk) [inline circuits/circom/updateStateTree.circom:257]: Better to derive 4 from a constant.

File: circuits/circom/updateStateTree.circom [hunk 3]
```diff
@@ -277,7 +280,7 @@ template UpdateStateTree(
     signal new_vote_options_tree_root;
     signal signature_verifier_valid;
 
-    var vote_options_max_leaves = 2 ** vote_options_tree_depth;
+    var vote_options_max_leaves = 5 ** vote_options_tree_depth;
```
Comment (kobigurk) [inline circuits/circom/updateStateTree.circom:283]: Better use a constant.

File: circuits/circom/updateStateTree.circom [hunk 4]
```diff
@@ -329,7 +332,9 @@ template UpdateStateTree(
     perform_checks_before_update.vote_options_tree_root <== vote_options_tree_root;
     perform_checks_before_update.vote_options_leaf_raw <== vote_options_leaf_raw;
     for (var i = 0; i < vote_options_tree_depth; i++) {
-        perform_checks_before_update.vote_options_tree_path_elements[i] <== vote_options_tree_path_elements[i];
+        for (var j = 0; j < 4; j++) {
```
Comment (kobigurk) [inline circuits/circom/updateStateTree.circom:335]: Better use a constant.

File: circuits/circom/updateStateTree.circom [hunk 5]
```diff
@@ -382,7 +387,7 @@ template UpdateStateTree(
     valid_state_leaf_index.in[0] <== decrypted_command_out[CMD_STATE_TREE_INDEX_IDX];
     valid_state_leaf_index.in[1] <== state_tree_max_leaf_index;
 
-    component valid_vote_options_leaf_index = LessEqThan(8);
+    component valid_vote_options_leaf_index = LessEqThan(32);
```
Comment (kobigurk) [inline circuits/circom/updateStateTree.circom:390]: Does this mean 2^32?

File: circuits/circom/trees/incrementalQuinTree.circom [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+template QuinSelector(choices) {
+    signal input in[choices];
+    signal input index;
+    signal output out;
+    
+    // Ensure that index < choices
+    component lessThan = LessThan(3);
+    lessThan.in[0] <== index;
+    lessThan.in[1] <== choices;
+    lessThan.out === 1;
```
Comment (kobigurk) [inline circuits/circom/trees/incrementalQuinTree.circom:36]: These are values provided by the coordinator, right? If yes, this condition is fine :)

File: circuits/circom/updateStateTree.circom (hunk 5 above)
Comment (weijiekoh) [reply circuits/circom/updateStateTree.circom:390]: I think so. `LessThan(n)` checks the `n`th bit of the output of a `Num2Bits`: https://github.com/iden3/circomlib/blob/master/circuits/comparators.circom#L88

Is `2 ** n` the maximum value for either of the input signals into `LessEqThan`?

File: circuits/circom/trees/incrementalQuinTree.circom (hunk 6 above)
Comment (weijiekoh) [reply circuits/circom/trees/incrementalQuinTree.circom:36]: Yup!

**Pull request merged**